### PR TITLE
[core.atomic] atomicStore for alias this

### PR DESCRIPTION
**Pull request by thewilsonator**
@kinke the corresponding LDC commit mentions some unit tests, but I presume that they are in the main LDC repo, but I can't fin the correspond PR/commit. Do you know where they are and should they be moved to druntime?

### CODE REVIEW

**Comment by kinke** [comment]
I couldn't find the unittest after a quick search either, maybe it has been removed in the meantime. Anyway, a simple test that works for LDC but not DMD due to this would be:
```D
unittest
{
    static struct S
    {
        int a;
        alias a this;
    }

    import core.atomic;
    S s;
    atomicStore(s, 123);
}
```

**Comment by thewilsonator** [comment]
Awesome. Thanks!

**Comment by thewilsonator** [comment]
Tests added.

File: src/core/atomic.d [hunk 1]
```diff
@@ -1204,4 +1204,17 @@ version (CoreUnittest)
...
+
+    unittest
+    {
+        static struct S
+        {
+            int a;
+            alias a this;
+        }
+
+        import core.atomic;
```
Comment (kinke) [inline src/core/atomic.d:1216]: no need ;)

Comment (thewilsonator) [reply src/core/atomic.d:1216]: Derp.

**Comment by Geod24** [comment]
Do we really need this implicit conversion magic though ? The argument is passed by `ref`, implicit conversions are *not* expected. I'd argue that anyone that wants to do this should specify `T` explicitly on the caller side.
If you really want implicit conversions, the signature should reflect it, e.g. T should be deduced to be the common acceptable type.

**Comment by thewilsonator** [comment]
Git blames https://github.com/dlang/druntime/pull/2745 for the comment ` // resolve implicit conversions` 
cc'ing @TurkeyMan

**Comment by TurkeyMan** [comment]
My refactor to simplify and untangle the implementation expects that all weird type wrangling is in the front-end wrappers. The `internal` functions deal only with one type.

"Implicit conversion" is for cases like `T == int` and `V == short`. Or `T == Base*` and `V == Derived*`.

The case in question with `alias this` feels weird to me. I'm not sure I call that an 'implicit conversion'... that's something else, and I don't know how deep that rabbit hole goes.
I think I'd rather make a patch to exclude this case (leading to compile error) rather than trying to handle it like this.

Should we add the constraint `is(V : T)` to the function?

**Comment by RazvanN7** [comment]
@Geod24 Since alias this is a language feature I don't see why we should not support it in this case. I wouldn't call it implicit conversion since the user explicitly defined the alias this.

I agree that a better solution would be to stress out the condition in the template constraint so that in case of screw ups the compiler doesn't point to the innards of `atomicStore`. 

@thewilsonator Please update the function signature so that we can get this in.

**Comment by thewilsonator** [comment]
> Please update the function signature so that we can get this in.

To what? `&& is(V : T)`?

**Comment by RazvanN7** [comment]
> To what? && is(V : T)?

Yes.

**Comment by thewilsonator** [comment]
Done..

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Thanks!

**Comment by thewilsonator** [comment]
Hmm, this probably needs a rebase

**Comment by RazvanN7** [comment]
Nope, the error is legit. You probably need to switch the types of the `is` expression: `is(V : T)`

**Comment by thewilsonator** [comment]
Sorry, seems I missed this. Suggestion applied. Lets see if this works.

**Comment by RazvanN7** [comment]
This doesn't seem to be going anywhere. Also, pondering upon it, I agree with @Geod24 : "Do we really need this implicit conversion magic though ? The argument is passed by ref, implicit conversions are not expected.". Feel free to reopen if you still want to pursue this.

**Pull request closed without merging**